diff --git a/bochs/cpu/rdrand.cc b/bochs/cpu/rdrand.cc
index cb26678fd..1438f43db 100644
--- a/bochs/cpu/rdrand.cc
+++ b/bochs/cpu/rdrand.cc
@@ -34,10 +34,12 @@ bool hw_rand_ready()
   return HW_RANDOM_GENERATOR_READY;
 }

+extern "C" Bit64u bochscpu_rand(unsigned);
+
 // provide a byte of data from Hardware Random Generator (TBD: implement as device)
 Bit8u hw_rand8()
 {
-  return rand() & 0xff;     // hack using std C rand() function
+  return bochscpu_rand(BX_CPU_ID) & 0xff;     // hack using std C rand() function
 }

 // provide a 16-bit of data from Hardware Random Generator (TBD: implement as device)
